adding edit_resolution method to support changing rollup_resolution of a datetime variable

mathiasbc opened the pull request:
This method is still not working, but I can't figure why. I am sending the data just like the UI is.

mathiasbc commented:
@jjdelc This is failing with:
```
pycrunch.lemonpy.ClientError: (<Response [400]>, 'https://app.crunch.io/api/datasets/b8446edded684f03a4388326266b80ff/variables/1c95095169314754924ccdf6ca879d8b/', {'status': '400 Bad Request', 'message': 'Invalid variable attributes: rollup_resolution'})
```
Strange, from my understanding the UI is doing a PATCH to the variable endpoint with a payload of `{'resolution_rollup': resolution}`

jjdelc commented:
The attribute is `resolution`, send it as `{resolution: new_resolution}`

jjdelc commented on scrunch/datasets.py:
```diff
@@ -2495,3 +2493,16 @@ def transformations(self):
+
+        payload = json.dumps({'rollup_resolution': resolution})
+        resp = self.resource.patch(payload)
+        self.dataset._reload_variables()
```
If you have `self.resource` to be the Variable shoji entity, you only need to do `self.resource.edit(resolution=new_resolution)` no need to do json.dumps.

mathiasbc replied:
yes that the original intention but since it was failing I decided to test other paths. Neither of the methods works for this attribute.

jjdelc commented on scrunch/datasets.py:
```diff
@@ -2495,3 +2493,16 @@ def transformations(self):
+        payload = json.dumps({'rollup_resolution': resolution})
+        resp = self.resource.patch(payload)
+        self.dataset._reload_variables()
+        return self.dataset[self.alias]
```
If you use the `.edit()` pycrunch method, then the resource's attribute will get automatically updated so no need to re-read the Variable instance or to do a full reload of all the variables (can be a expensive GET). You could just return `self`.

mathiasbc replied:
👍

mathiasbc commented:
@jjdelc: resolution is a different attribute and in this case we don't intend to change that. @jamesrkg noticed that the UI only changes `resolution_rollup ` and `resolution` remains the same. The payload sent by the UI is exactly as I pasted above.

jjdelc commented:
Ah, I see. The attribute is `rollup_resolution`, "rollup" is first. http://docs.crunch.io/endpoint-reference/endpoint-variable.html

mathiasbc commented:
@jjdelc: Ah sorry that was my mistake in the comment, but the code actually had the correct name. I made some updates, still I get:
pycrunch.lemonpy.ClientError: (<Response [400]>, 'https://app.crunch.io/api/datasets/b8446edded684f03a4388326266b80ff/variables/1c95095169314754924ccdf6ca879d8b/', {'status': '400 Bad Request', 'message': 'Invalid variable attributes: rollup_resolution'})

mathiasbc commented:
@jjdelc @jamesrkg @xbito This should be good to go now. Changed a few things. Fernando helped me realized we need to change the value inside the `view` attribute and not the external `rollup_resolution`. Renamed the method to `edit_resolution` to avoid the confusion with the variable attribute `rollup_resolution`. 

Note: for this method to work the user needs to grab the Dataset editor's lock. `get_dataset(ds_id, editor=True)`

jjdelc commented on scrunch/datasets.py:
```diff
@@ -2495,3 +2493,18 @@ def transformations(self):
+        """
+        assert self.type == 'datetime', 'Method only allowed for datetime variables'
+        self.dataset._validate_vartypes(self.type, resolution=resolution)
+        self.resource.edit(view=dict(rollup_resolution=resolution))
```
This is wrong, note that here you will overwrite all the other attributes in the variable's view attribute. You need to read the current dict, then change the value an send that as the new dict. Sending a new dict will overwrite the full thing.
